### PR TITLE
♻️ change call-service row to button row

### PR DESCRIPTION
**Pull request by iantrich**
## Proposed change

Rename `call-service` special row type to `button`. Will allow us to better align between row/card/element. Non-breaking change as `call-service` row now just extends `button` row

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

image
```yaml
type: entities
entities:
  - entity: light.bed_light
  - type: button
    service: light.toggle
    name: Light Toggle Button
    action_name: Button
    service_data:
      entity_id: light.bed_light
  - type: call-service
    service: light.toggle
    name: Light Toggle Call-Service
    action_name: Call-Service
    service_data:
      entity_id: light.bed_light
```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12545

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
Nice, we should have scene and script row extend this as well.

File: src/panels/lovelace/entity-rows/types.ts [hunk 1]
```diff
@@ -28,8 +28,8 @@ export interface WeblinkConfig {
   icon?: string;
   url: string;
 }
-export interface CallServiceConfig extends EntityConfig {
-  type: "call-service";
+export interface ButtonRowConfig extends EntityConfig {
+  type: "button" | "call-service";
   action_name?: string;
```
Comment (bramkragten) [inline src/panels/lovelace/entity-rows/types.ts]: If we are creating a new type, shouldn't we align the option with the button card? So add the default actions config instead of this?

Comment (bramkragten) [reply src/panels/lovelace/entity-rows/types.ts]: We could have the `call-service` row config map to `button` config:
```ts
{tap_action: { action: "call-service", service: config.service, service_data: service_data}  }
```

Comment (iantrich) [reply src/panels/lovelace/entity-rows/types.ts]: yes, that's a good idea

Comment (iantrich) [reply src/panels/lovelace/entity-rows/types.ts]: So was about to start on this and realized that we already have actions for rows...but they are for the icon/name already. So we'd need to differentiate that these actions are for the button vs the "row". Thoughts?
```yaml
button_action:
  tap_action:
  ...
```

Comment (bramkragten) [reply src/panels/lovelace/entity-rows/types.ts]: That sucks, getting complicated now.
I think that is our only option?

Comment (zsarnett) [reply src/panels/lovelace/entity-rows/types.ts]: Why differentiate them? If the Tap Action for either row then the Row and the button do the same action. I don't see the need to do two different user defined actions. Tapping the Row should show more info and the button should do the action

Comment (bramkragten) [reply src/panels/lovelace/entity-rows/types.ts]: Yeah that's how a scene row works for instance.

Comment (zsarnett) [reply src/panels/lovelace/entity-rows/types.ts]: Yea. I think that is a better option than confusing the configuration with the proposed config above

Comment (iantrich) [reply src/panels/lovelace/entity-rows/types.ts]: I agree, somewhat. We can make this work with existing actions, but actions on the icon/name are set to 'none'

Comment (zsarnett) [reply src/panels/lovelace/entity-rows/types.ts]: I feel like its more of an aesthetics thing for users. It's not intuitive that clicking the icon or name would perform an action. I would think it would bring up more info and the button is the only thing that performs the action.

Comment (iantrich) [reply src/panels/lovelace/entity-rows/types.ts]: I spoke to this without actually looking at the code base, there are already no actions on the icon/name so bram's initial suggestion was perfect all along :)

File: src/panels/lovelace/special-rows/hui-button-row.ts [hunk 2]
```diff
@@ -25,7 +28,15 @@ export class HuiButtonRow extends LitElement implements LovelaceRow {
       throw new Error("Error in card configuration.");
     }
 
-    this._config = { icon: "hass:remote", ...config };
+    this._config = {
+      icon: "hass:remote",
+      tap_action: {
+        action: "call-service",
+        service: config.service,
```
Comment (bramkragten) [inline src/panels/lovelace/special-rows/hui-button-row.ts]: This should not be in `hui-button-row` but in `hui-call-service-row`

File: src/panels/lovelace/entity-rows/types.ts [hunk 3]
```diff
@@ -28,7 +29,7 @@ export interface WeblinkConfig {
   icon?: string;
   url: string;
 }
-export interface ButtonRowConfig extends EntityConfig {
+export interface ButtonRowConfig extends EntitiesCardEntityConfig {
   type: "button" | "call-service";
   action_name?: string;
   service: string;
```
Comment (bramkragten) [inline src/panels/lovelace/entity-rows/types.ts]: Remove this from `ButtonRowConfig`

File: src/panels/lovelace/entity-rows/types.ts [hunk 4]
```diff
@@ -28,8 +29,8 @@ export interface WeblinkConfig {
   icon?: string;
   url: string;
 }
-export interface CallServiceConfig extends EntityConfig {
-  type: "call-service";
+export interface ButtonRowConfig extends EntitiesCardEntityConfig {
+  type: "button" | "call-service";
   action_name?: string;
   service: string;
   service_data?: { [key: string]: any };
```
Comment (bramkragten) [inline src/panels/lovelace/entity-rows/types.ts]: Remove this from `ButtonRowConfig`

File: src/panels/lovelace/special-rows/hui-button-row.ts (hunk 2 above)
Comment (iantrich) [reply src/panels/lovelace/special-rows/hui-button-row.ts]: So you want hui-call-service-row to implement its own setConfig? There will be clashes between the two as the types don't overlap.

Comment (bramkragten) [reply src/panels/lovelace/special-rows/hui-button-row.ts]: That doesn't matter does it? `call-service-row` will be an individual element that uses `button-row` to be rendered?

File: src/panels/lovelace/special-rows/hui-call-service-row.ts [hunk 5]
```diff
@@ -4,80 +4,63 @@ import {
...
+  LovelaceRow,
+  CallServiceRowConfig,
+  ButtonRowConfig,
+} from "../entity-rows/types";
 import { HomeAssistant } from "../../../types";
+import { createRowElement } from "../create-element/create-row-element";
 
 @customElement("hui-call-service-row")
-class HuiCallServiceRow extends LitElement implements LovelaceRow {
+export class HuiCallServiceRow extends LitElement implements LovelaceRow {
```
Comment (bramkragten) [inline src/panels/lovelace/special-rows/hui-call-service-row.ts]: Can we extend `HuiButtonRow` and only override the setConfig function?

File: src/panels/lovelace/special-rows/hui-button-row.ts [hunk 6]
```diff
@@ -21,7 +21,7 @@ import { handleAction } from "../common/handle-action";
 @customElement("hui-button-row")
 export class HuiButtonRow extends LitElement implements LovelaceRow {
   public hass?: HomeAssistant;
-  @property() private _config?: ButtonRowConfig;
+  @property() protected _config?: ButtonRowConfig;
```
Comment (zsarnett) [inline src/panels/lovelace/special-rows/hui-button-row.ts:24]: I dont think you need to do this. Should be able just call `super.setConfig(config)`

Comment (iantrich) [reply src/panels/lovelace/special-rows/hui-button-row.ts]: good call

File: src/panels/lovelace/special-rows/hui-call-service-row.ts [hunk 7]
```diff
@@ -1,83 +1,34 @@
...
+import { HuiButtonRow } from "./hui-button-row";
 
 @customElement("hui-call-service-row")
-class HuiCallServiceRow extends LitElement implements LovelaceRow {
-  public hass?: HomeAssistant;
-
-  @property() private _config?: CallServiceConfig;
+export class HuiCallServiceRow extends HuiButtonRow {
+  public setConfig(config: ButtonRowConfig): void {
+    const callServiceConfig: any = config;
```
Comment (bramkragten) [inline src/panels/lovelace/special-rows/hui-call-service-row.ts:9]: Why should keep the interface for `CallServiceConfig`

File: src/panels/lovelace/special-rows/hui-call-service-row.ts [hunk 8]
```diff
@@ -1,83 +1,34 @@
...
+import { ButtonRowConfig } from "../entity-rows/types";
+import { HuiButtonRow } from "./hui-button-row";
 
 @customElement("hui-call-service-row")
-class HuiCallServiceRow extends LitElement implements LovelaceRow {
-  public hass?: HomeAssistant;
-
-  @property() private _config?: CallServiceConfig;
+export class HuiCallServiceRow extends HuiButtonRow {
+  public setConfig(config: ButtonRowConfig): void {
```
Comment (bramkragten) [inline src/panels/lovelace/special-rows/hui-call-service-row.ts:8]:
```suggestion
  public setConfig(config: CallServiceConfig): void {
```

**Review by bramkragten — APPROVED** [review]

File: src/panels/lovelace/special-rows/hui-call-service-row.ts (hunk 8 above)
Comment (iantrich) [reply src/panels/lovelace/special-rows/hui-call-service-row.ts]: Can't change the signature of setConfig

Comment (bramkragten) [reply src/panels/lovelace/special-rows/hui-call-service-row.ts]: Yeah already "fixed" it...

Comment (zsarnett) [reply src/panels/lovelace/special-rows/hui-call-service-row.ts]: I think the best way would be

```ts
const buttonRowConfig = {...config, type: 'button'};

super.setConfig(buttonRowConfig);
```

Similar to how the Sensor card does for entity card extending

File: src/panels/lovelace/entity-rows/types.ts [hunk 9]
```diff
@@ -62,9 +62,9 @@ export type LovelaceRowConfig =
   | WeblinkConfig
   | ButtonsRowConfig
   | ConditionalRowConfig
-  | ButtonRowConfig
```
Comment (iantrich) [inline src/panels/lovelace/entity-rows/types.ts]: why did you remove this?

Comment (bramkragten) [reply src/panels/lovelace/entity-rows/types.ts]: Freaking merge conflict
